corehq/apps/geospatial/reports.py

```diff
@@ -90,76 +128,243 @@ def _get_geo_location(case):
+        #     {
+        #         'key': 't0'
+        #         'doc_count': 1,
+        #         'bucket_cases': {
+        #             'bounds': {
+        #                 'bottom_right': {
+        #                     'lat': 4.912349972873926,
+        #                     'lon': 52.374080987647176,
+        #                 },
+        #                 'top_left': {
+        #                     'lat': 4.912349972873926,
+        #                     'lon': 52.374080987647176,
+        #                 },
+        #             },
+        #         },
+        #     }
+        bounds = bucket[BUCKET_CASES_AGG]['bounds']
+
+        # If there is only one case in the bucket, then the bounds will
+        # be the geo_point of that case, and top_left and bottom_right
+        # will be the same.
+        #
+        # If they are the same, then the geo_bounding_box filter will
+        # error in Elasticsearch 5.6. (This is not a problem in
+        # Elasticsearch 8+, where we can use the bucket key, which is
+        # its geohash, to select the cases in the bucket.)
+        #
+        # So we shift the top left and bottom right by 0.000_01 degrees,
+        # or roughly 1 metre.
+        if bounds['top_left']['lat'] == bounds['bottom_right']['lat']:
+            bounds['top_left']['lat'] += 0.000_01
+            bounds['bottom_right']['lat'] -= 0.000_01
+        if bounds['top_left']['lon'] == bounds['bottom_right']['lon']:
+            bounds['top_left']['lon'] -= 0.000_01
+            bounds['bottom_right']['lon'] += 0.000_01
 
-        # Apply geohash grid aggregation
+        query = super()._build_query()  # `super()` so as not to filter
+        filters_ = [filters.geo_bounding_box(
+            field=PROPERTY_GEOPOINT_VALUE,
+            top_left=bounds['top_left'],
+            bottom_right=bounds['bottom_right'],
+        )]
+        if self.request.GET.get('features'):
+            features = json.loads(self.request.GET['features'])
+            features_filter = self._get_filter_for_features(features)
+            filters_.append(features_filter)
+        query = self._filter_query(query, filters_)
+        es_results = query.run().raw
+
+        cases = []
+        for row in es_results['hits'].get('hits', []):
+            display = CaseDisplayES(
+                self.get_case(row), self.timezone, self.individual
+            )
+            case = wrap_case_search_hit(row)
+            coordinates = self._get_geo_location(case)
+            cases.append([
+                display.case_id,
+                coordinates,
+                display.case_link
+            ])
+        return cases
+
+    @quickcache(['self.domain', 'self.shared_pagination_GET_params'], timeout=15 * 60)
+    def _get_buckets(self):
+        query = self._build_query()
+        query = self._aggregate_query(query)
+        es_results = query.run().raw
+        if es_results is None:
+            return []
+        return (
```

Comment:
Just a comment: I would love to see datastructures/python objects being used instead of raw dictionaries. Would that be feasible? (Of course, not for this PR)